File: src/lib/dal.ts
/**
 * Data Access Layer — every SQL query in the application lives here.
 *
 * If a table is renamed, a column changes, or a query needs updating,
 * this is the ONLY file you need to touch.
 */

import { query } from "./db.server";

// ---------------------------------------------------------------------------
// Table names — single source of truth
// ---------------------------------------------------------------------------

const T = {
  CUSTOM_SCAN: "custom_scan",
  LIVE_NEWS: "live_news",
  CA_DIVIDENDS: "corporate_actions_dividends",
  CA_BONUS: "corporate_actions_bonus",
  CA_SPLITS: "corporate_actions_splits",
  CA_RIGHTS: "corporate_actions_rights",
  CA_BUYBACKS: "corporate_actions_buybacks",
  CA_QUARTERLY_RESULTS: "corporate_actions_quarterly_results",
} as const;

/** Maps a corporate-action URL key to its database table name. */
export const CORPORATE_ACTION_TABLES: Record<string, string> = {
  dividends: T.CA_DIVIDENDS,
  bonus: T.CA_BONUS,
  splits: T.CA_SPLITS,
  rights: T.CA_RIGHTS,
  buybacks: T.CA_BUYBACKS,
  "quarterly-results": T.CA_QUARTERLY_RESULTS,
};

// ---------------------------------------------------------------------------
// Reusable column lists
// ---------------------------------------------------------------------------

const STOCK_SUMMARY_COLS = "isin, sym, disp_sym, ltp, pperchange, mcapclass";

const SCREENER_COLS =
  "isin, sym, disp_sym, ltp, pperchange, mcap, mcapclass, pe, pb, div_yeild, roce, roe, eps, net_profit_margin, price_perchng_1year, volume";

const PEER_COLS =
  "isin, sym, disp_sym, ltp, pperchange, mcap, pe, div_yeild, roce, roe, eps, pb, net_profit_margin, volume";

const NEWS_FULL_COLS =
  "article_id, title, text, overall_sentiment, category, sub_category, publish_date, stock_name, isin_code, display_symbol, article_slug";

const NEWS_HEADLINE_COLS =
  "article_id, title, overall_sentiment, category, sub_category, publish_date";

const CORPORATE_ACTION_COLS =
  "row_hash, isin, sym, disp_sym, exch, inst, seg, seosym, ltp, volume, pchange, pperchange, act_type, ann_date, ann_ltp, div_type, ex_date, note, rec_date, rmk, fetched_at";

// ---------------------------------------------------------------------------
// custom_scan queries
// ---------------------------------------------------------------------------

/** Full stock row by ISIN (used on the stock detail page). */
export async function getStockByIsin(isin: string) {
  const rows = await query<Record<string, string | number | null>>(
    `SELECT * FROM ${T.CUSTOM_SCAN} WHERE isin = $1 LIMIT 1`,
    [isin],
  );
  return rows[0] ?? null;
}

/** Total number of rows in custom_scan. */
export async function getCustomScanCount(): Promise<number> {
  const rows = await query<{ count: string }>(
    `SELECT COUNT(*)::text AS count FROM ${T.CUSTOM_SCAN}`,
  );
  return Number(rows[0]?.count ?? 0);
}

/** Top gainer by percentage change. */
export async function getTopGainer() {
  const rows = await query<Record<string, unknown>>(
    `SELECT ${STOCK_SUMMARY_COLS}
     FROM ${T.CUSTOM_SCAN}
     WHERE pperchange IS NOT NULL
     ORDER BY pperchange DESC NULLS LAST LIMIT 1`,
  );
  return rows[0] ?? null;
}

/** Top loser by percentage change. */
export async function getTopLoser() {
  const rows = await query<Record<string, unknown>>(
    `SELECT ${STOCK_SUMMARY_COLS}
     FROM ${T.CUSTOM_SCAN}
     WHERE pperchange IS NOT NULL
     ORDER BY pperchange ASC NULLS LAST LIMIT 1`,
  );
  return rows[0] ?? null;
}

/** Top 10 stocks by market cap. */
export async function getPopularStocks() {
  return query<Record<string, unknown>>(
    `SELECT ${STOCK_SUMMARY_COLS}
     FROM ${T.CUSTOM_SCAN}
     ORDER BY mcap DESC NULLS LAST LIMIT 10`,
  );
}

/** Search stocks by symbol (ILIKE match). */
export async function searchStocks(q: string) {
  const like = `%${q}%`;
  return query<Record<string, unknown>>(
    `SELECT ${STOCK_SUMMARY_COLS}
     FROM ${T.CUSTOM_SCAN}
     WHERE disp_sym ILIKE $1 OR sym ILIKE $1
     ORDER BY
       CASE WHEN sym ILIKE $2 THEN 0
            WHEN disp_sym ILIKE $2 THEN 1
            ELSE 2 END,
       mcap DESC NULLS LAST
     LIMIT 10`,
    [like, `${q}%`],
  );
}

/** Look up the mcapclass for a given ISIN. */
export async function getMcapclassByIsin(isin: string) {
  const rows = await query<{ mcapclass: string | null }>(
    `SELECT mcapclass FROM ${T.CUSTOM_SCAN} WHERE isin = $1 LIMIT 1`,
    [isin],
  );
  return rows[0] ?? null;
}

/** Peer stocks in the same mcapclass (top 10 by market cap). */
export async function getPeerStocks(mcapclass: string) {
  return query<Record<string, unknown>>(
    `SELECT ${PEER_COLS} FROM ${T.CUSTOM_SCAN}
     WHERE mcapclass = $1
     ORDER BY mcap DESC NULLS LAST LIMIT 10`,
    [mcapclass],
  );
}

/** Single stock with peer-comparison columns. */
export async function getStockWithPeerCols(isin: string) {
  const rows = await query<Record<string, unknown>>(
    `SELECT ${PEER_COLS} FROM ${T.CUSTOM_SCAN} WHERE isin = $1 LIMIT 1`,
    [isin],
  );
  return rows[0] ?? null;
}

// --- Screener ---------------------------------------------------------------

const SORTABLE = new Set([
  "mcap",
  "ltp",
  "pperchange",
  "pe",
  "pb",
  "div_yeild",
  "roce",
  "roe",
  "eps",
  "net_profit_margin",
  "price_perchng_1year",
  "volume",
]);

export type ScreenerFilters = {
  mcapclass?: string[];
  peMin?: number;
  peMax?: number;
  roceMin?: number;
  roeMin?: number;
  divYieldMin?: number;
  sort?: string;
  order?: "asc" | "desc";
  page?: number;
  limit?: number;
};

/** Paginated, filtered screener results together with the total count. */
export async function getScreenerResults(filters: ScreenerFilters) {
  const where: string[] = [];
  const args: unknown[] = [];
  const push = (clause: string, val: unknown) => {
    args.push(val);
    where.push(clause.replace("?", `$${args.length}`));
  };

  if (filters.mcapclass?.length) {
    args.push(filters.mcapclass);
    where.push(`mcapclass = ANY($${args.length})`);
  }
  if (filters.peMin != null) push(`pe >= ?`, filters.peMin);
  if (filters.peMax != null) push(`pe <= ?`, filters.peMax);
  if (filters.roceMin != null) push(`roce >= ?`, filters.roceMin);
  if (filters.roeMin != null) push(`roe >= ?`, filters.roeMin);
  if (filters.divYieldMin != null) push(`div_yeild >= ?`, filters.divYieldMin);

  const sortRaw = filters.sort ?? "mcap";
  const sort = SORTABLE.has(sortRaw) ? sortRaw : "mcap";
  const order = filters.order === "asc" ? "ASC" : "DESC";

  const page = Math.max(1, filters.page ?? 1);
  const limit = Math.min(200, Math.max(1, filters.limit ?? 50));
  const offset = (page - 1) * limit;

  const whereSql = where.length ? `WHERE ${where.join(" AND ")}` : "";

  const countRows = await query<{ c: string | number }>(
    `SELECT COUNT(*)::int AS c FROM ${T.CUSTOM_SCAN} ${whereSql}`,
    args,
  );
  const total = Number(countRows[0]?.c ?? 0);

  const rows = await query<Record<string, unknown>>(
    `SELECT ${SCREENER_COLS} FROM ${T.CUSTOM_SCAN} ${whereSql} ORDER BY ${sort} ${order} NULLS LAST LIMIT ${limit} OFFSET ${offset}`,
    args,
  );

  return { rows, total, page, limit };
}

/** Distinct mcapclass values and total stock count (screener filter facets). */
export async function getScreenerFacets() {
  const classes = await query<{ mcapclass: string | null }>(
    `SELECT DISTINCT mcapclass FROM ${T.CUSTOM_SCAN}
     WHERE mcapclass IS NOT NULL
     ORDER BY mcapclass`,
  );
  const totalRows = await query<{ c: string | number }>(
    `SELECT COUNT(*)::int AS c FROM ${T.CUSTOM_SCAN}`,
  );

  return {
    mcapclasses: classes.map((r) => r.mcapclass).filter(Boolean) as string[],
    total: Number(totalRows[0]?.c ?? 0),
  };
}

// ---------------------------------------------------------------------------
// live_news queries
// ---------------------------------------------------------------------------

/** Latest news articles (global feed). */
export async function getLatestNews(limit: number) {
  return query<Record<string, unknown>>(
    `SELECT ${NEWS_FULL_COLS}
     FROM ${T.LIVE_NEWS}
     WHERE publish_date IS NOT NULL
     ORDER BY publish_date DESC
     LIMIT $1`,
    [limit],
  );
}

/** News headlines for a specific stock by ISIN. */
export async function getNewsByIsin(isin: string) {
  return query<Record<string, unknown>>(
    `SELECT ${NEWS_HEADLINE_COLS}
     FROM ${T.LIVE_NEWS}
     WHERE isin_code = $1
     ORDER BY publish_date DESC NULLS LAST
     LIMIT 10`,
    [isin],
  );
}

// ---------------------------------------------------------------------------
// corporate_actions queries
// ---------------------------------------------------------------------------

/**
 * Rows for a specific corporate action type.
 * Returns `null` if the action key is unknown.
 */
export async function getCorporateActionRows(actionKey: string, isin: string) {
  const table = CORPORATE_ACTION_TABLES[actionKey];
  if (!table) return null;

  return query<Record<string, unknown>>(
    `SELECT ${CORPORATE_ACTION_COLS}
     FROM ${table}
     WHERE isin = $1
     ORDER BY ex_date DESC NULLS LAST`,
    [isin],
  );
}

/** Row counts across all corporate action tables for a given ISIN. */
export async function getCorporateActionCounts(isin: string) {
  const entries = await Promise.all(
    Object.entries(CORPORATE_ACTION_TABLES).map(async ([key, table]) => {
      const rows = await query<{ c: string | number }>(
        `SELECT COUNT(*)::int AS c FROM ${table} WHERE isin = $1`,
        [isin],
      );
      return [key, Number(rows[0]?.c ?? 0)] as const;
    }),
  );
  return Object.fromEntries(entries);
}

File: Backend/dal.py
"""
Data Access Layer — every table name, column list, and SQL template for the
Backend ingestion pipeline lives here.

If a table is renamed or a column changes, this is the ONLY file you need to
touch.  The individual ingest scripts import from here instead of hard-coding
table names / SQL.
"""

import os
import psycopg2
from psycopg2.extras import execute_values
from dotenv import load_dotenv

# ---------------------------------------------------------------------------
# Environment
# ---------------------------------------------------------------------------

_PROJECT_ROOT = os.path.dirname(os.path.abspath(__file__))
_local_env = os.path.join(_PROJECT_ROOT, ".env")
_parent_env = os.path.join(os.path.dirname(_PROJECT_ROOT), ".env")
if os.path.exists(_local_env):
    load_dotenv(_local_env)
else:
    load_dotenv(_parent_env)

DATABASE_URL = os.getenv("DATABASE_URL")

# ---------------------------------------------------------------------------
# Table names — single source of truth
# ---------------------------------------------------------------------------

T_CUSTOM_SCAN = "custom_scan"
T_LIVE_NEWS = "live_news"
T_INGESTION_METADATA = "ingestion_metadata"

T_CA_DIVIDENDS = "corporate_actions_dividends"
T_CA_BONUS = "corporate_actions_bonus"
T_CA_SPLITS = "corporate_actions_splits"
T_CA_RIGHTS = "corporate_actions_rights"
T_CA_BUYBACKS = "corporate_actions_buybacks"
T_CA_QUARTERLY_RESULTS = "corporate_actions_quarterly_results"

# Ordered list of all corporate-action tables (used for init_tables, etc.)
CORPORATE_ACTION_TABLES = [
    T_CA_QUARTERLY_RESULTS,
    T_CA_BONUS,
    T_CA_DIVIDENDS,
    T_CA_SPLITS,
    T_CA_RIGHTS,
    T_CA_BUYBACKS,
]

# Maps API action-type strings to their table names
ACTION_TYPE_TO_TABLE = {
    "QUARTERLY RESULT ANNOUNCEMENT": T_CA_QUARTERLY_RESULTS,
    "BONUS": T_CA_BONUS,
    "DIVIDEND": T_CA_DIVIDENDS,
    "Splits": T_CA_SPLITS,
    "Rights": T_CA_RIGHTS,
    "Buyback": T_CA_BUYBACKS,
}

# ---------------------------------------------------------------------------
# Connection helper
# ---------------------------------------------------------------------------

def get_connection():
    """Return a new psycopg2 connection using DATABASE_URL."""
    if not DATABASE_URL:
        raise ValueError("DATABASE_URL is not set in the environment or .env file.")
    return psycopg2.connect(DATABASE_URL)


# ---------------------------------------------------------------------------
# Ingestion metadata queries
# ---------------------------------------------------------------------------

def init_metadata_table():
    """CREATE TABLE IF NOT EXISTS for ingestion_metadata."""
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(f"""
                CREATE TABLE IF NOT EXISTS {T_INGESTION_METADATA} (
                    script_name VARCHAR(100) PRIMARY KEY,
                    last_run_date DATE NOT NULL,
                    updated_at TIMESTAMP WITH TIME ZONE NOT NULL
                );
            """)
        conn.commit()
    except Exception:
        conn.rollback()
        raise
    finally:
        conn.close()


def get_last_run_date(script_name):
    """Return the last_run_date for *script_name*, or None if never run."""
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                f"SELECT last_run_date FROM {T_INGESTION_METADATA} WHERE script_name = %s",
                (script_name,),
            )
            row = cur.fetchone()
            return row[0] if row else None
    finally:
        conn.close()


def update_last_run_date(script_name, run_date):
    """Upsert the last_run_date for *script_name*."""
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(f"""
                INSERT INTO {T_INGESTION_METADATA} (script_name, last_run_date, updated_at)
                VALUES (%s, %s, NOW())
                ON CONFLICT (script_name) DO UPDATE SET
                    last_run_date = EXCLUDED.last_run_date,
                    updated_at = EXCLUDED.updated_at
            """, (script_name, run_date))
        conn.commit()
    finally:
        conn.close()


# ---------------------------------------------------------------------------
# Corporate-actions table init + queries
# ---------------------------------------------------------------------------

def init_corporate_action_tables(logger):
    """CREATE TABLE IF NOT EXISTS for every corporate-action table."""
    logger.info("Initializing corporate actions tables...")
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            for table in CORPORATE_ACTION_TABLES:
                cur.execute(f"""
                    CREATE TABLE IF NOT EXISTS {table} (
                        row_hash VARCHAR(64) PRIMARY KEY,
                        isin VARCHAR(20) NOT NULL,
                        sym VARCHAR(20) NOT NULL,
                        disp_sym VARCHAR(100),
                        exch VARCHAR(10),
                        inst VARCHAR(20),
                        seg VARCHAR(10),
                        seosym VARCHAR(100),
                        ltp DOUBLE PRECISION,
                        volume BIGINT,
                        pchange DOUBLE PRECISION,
                        pperchange DOUBLE PRECISION,
                        act_type VARCHAR(100) NOT NULL,
                        ann_date DATE,
                        ann_ltp DOUBLE PRECISION,
                        div_type VARCHAR(50),
                        ex_date DATE,
                        note TEXT,
                        rec_date DATE,
                        rmk TEXT,
                        fetched_at TIMESTAMP WITH TIME ZONE NOT NULL
                    );
                """)
        conn.commit()
        logger.info("Corporate actions tables initialized successfully.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to initialize tables: {e}")
        raise
    finally:
        conn.close()


def load_existing_hashes(conn, table_name, logger):
    """Return the set of row_hash values already in *table_name*."""
    existing = set()
    with conn.cursor() as cur:
        try:
            cur.execute(f"SELECT row_hash FROM {table_name};")
            for row in cur.fetchall():
                existing.add(row[0])
        except Exception as e:
            logger.warning(f"Could not load existing records for {table_name}: {e}")
    return existing


def insert_corporate_action_records(table_name, records, logger):
    """Upsert corporate-action tuples into *table_name*."""
    if not records:
        return

    sql = f"""
        INSERT INTO {table_name} (
            row_hash, isin, sym, disp_sym, exch, inst, seg, seosym,
            ltp, volume, pchange, pperchange,
            act_type, ann_date, ann_ltp, div_type, ex_date, note, rec_date, rmk,
            fetched_at
        ) VALUES %s
        ON CONFLICT (row_hash) DO UPDATE SET
            isin = EXCLUDED.isin,
            sym = EXCLUDED.sym,
            disp_sym = EXCLUDED.disp_sym,
            exch = EXCLUDED.exch,
            inst = EXCLUDED.inst,
            seg = EXCLUDED.seg,
            seosym = EXCLUDED.seosym,
            ltp = EXCLUDED.ltp,
            volume = EXCLUDED.volume,
            pchange = EXCLUDED.pchange,
            pperchange = EXCLUDED.pperchange,
            ann_date = EXCLUDED.ann_date,
            ann_ltp = EXCLUDED.ann_ltp,
            div_type = EXCLUDED.div_type,
            ex_date = EXCLUDED.ex_date,
            note = EXCLUDED.note,
            rec_date = EXCLUDED.rec_date,
            rmk = EXCLUDED.rmk,
            fetched_at = EXCLUDED.fetched_at;
    """

    conn = get_connection()
    try:
        with conn.cursor() as cur:
            execute_values(cur, sql, records)
        conn.commit()
        logger.info(f"Inserted {len(records)} records into '{table_name}'.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to insert records into '{table_name}': {e}")
        raise
    finally:
        conn.close()


# ---------------------------------------------------------------------------
# custom_scan table init + queries
# ---------------------------------------------------------------------------

def init_custom_scan_table(logger):
    """CREATE TABLE IF NOT EXISTS for custom_scan."""
    logger.info("Initializing custom_scan table...")
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(f"""
            CREATE TABLE IF NOT EXISTS {T_CUSTOM_SCAN} (
                isin VARCHAR(20) NOT NULL,
                sym VARCHAR(20) NOT NULL,
                disp_sym VARCHAR(100),
                exch VARCHAR(10),
                inst VARCHAR(20),
                seg VARCHAR(10),
                seosym VARCHAR(100),
                mcap DOUBLE PRECISION,
                mcapclass VARCHAR(50),
                pe DOUBLE PRECISION,
                div_yeild DOUBLE PRECISION,
                revenue DOUBLE PRECISION,
                year_1_revenue_growth DOUBLE PRECISION,
                net_profit_margin DOUBLE PRECISION,
                yoy_last_qtrly_profit_growth DOUBLE PRECISION,
                ebidta_margin DOUBLE PRECISION,
                volume BIGINT,
                price_perchng_1year DOUBLE PRECISION,
                price_perchng_3year DOUBLE PRECISION,
                price_perchng_5year DOUBLE PRECISION,
                ind_pe DOUBLE PRECISION,
                pb DOUBLE PRECISION,
                eps DOUBLE PRECISION,
                day_sma_50_current_candle DOUBLE PRECISION,
                day_sma_200_current_candle DOUBLE PRECISION,
                day_rsi_14_current_candle DOUBLE PRECISION,
                roce DOUBLE PRECISION,
                ltp DOUBLE PRECISION,
                roe DOUBLE PRECISION,
                rt_away_from_5_year_high DOUBLE PRECISION,
                rt_away_from_1_month_high DOUBLE PRECISION,
                high_5yr DOUBLE PRECISION,
                high_3yr DOUBLE PRECISION,
                high_1yr DOUBLE PRECISION,
                high_1wk DOUBLE PRECISION,
                price_perchng_1mon DOUBLE PRECISION,
                price_perchng_1week DOUBLE PRECISION,
                price_perchng_3mon DOUBLE PRECISION,
                yearly_earning_per_share DOUBLE PRECISION,
                ocf_growth_on_yr DOUBLE PRECISION,
                year_1_cagr_eps_growth DOUBLE PRECISION,
                net_change_in_cash DOUBLE PRECISION,
                free_cash_flow DOUBLE PRECISION,
                price_perchng_2week DOUBLE PRECISION,
                day_bb_upper_sub_bb_lower DOUBLE PRECISION,
                day_atr_14_current_candle_mul_2 DOUBLE PRECISION,
                min_5_high_current_candle DOUBLE PRECISION,
                min_15_high_current_candle DOUBLE PRECISION,
                min_5_ema_50_current_candle DOUBLE PRECISION,
                min_15_ema_50_current_candle DOUBLE PRECISION,
                min_15_sma_100_current_candle DOUBLE PRECISION,
                open DOUBLE PRECISION,
                bc_close DOUBLE PRECISION,
                rmp DOUBLE PRECISION,
                pledge_benefit DOUBLE PRECISION,
                lot_size INTEGER,
                low_1yr DOUBLE PRECISION,
                multiplier INTEGER,
                pchange DOUBLE PRECISION,
                pperchange DOUBLE PRECISION,
                tick_size DOUBLE PRECISION,
                fetched_at TIMESTAMP WITH TIME ZONE NOT NULL,
                PRIMARY KEY (isin)
            );
            """)
        conn.commit()
        logger.info("custom_scan table initialized successfully.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to initialize table: {e}")
        raise e
    finally:
        conn.close()


def load_existing_isins(conn, logger):
    """Return the set of ISINs already in custom_scan."""
    existing = set()
    with conn.cursor() as cur:
        try:
            cur.execute(f"SELECT isin FROM {T_CUSTOM_SCAN};")
            for row in cur.fetchall():
                existing.add(row[0])
        except Exception as e:
            logger.warning(f"Could not load existing records: {e}")
    return existing


def upsert_custom_scan_records(records, logger):
    """Upsert custom_scan tuples."""
    if not records:
        return

    sql = f"""
        INSERT INTO {T_CUSTOM_SCAN} (
            isin, sym, disp_sym, exch, inst, seg, seosym, mcap, mcapclass, pe, div_yeild, revenue,
            year_1_revenue_growth, net_profit_margin, yoy_last_qtrly_profit_growth, ebidta_margin, volume,
            price_perchng_1year, price_perchng_3year, price_perchng_5year, ind_pe, pb, eps,
            day_sma_50_current_candle, day_sma_200_current_candle, day_rsi_14_current_candle, roce, ltp, roe,
            rt_away_from_5_year_high, rt_away_from_1_month_high, high_5yr, high_3yr, high_1yr, high_1wk,
            price_perchng_1mon, price_perchng_1week, price_perchng_3mon, yearly_earning_per_share,
            ocf_growth_on_yr, year_1_cagr_eps_growth, net_change_in_cash, free_cash_flow, price_perchng_2week,
            day_bb_upper_sub_bb_lower, day_atr_14_current_candle_mul_2, min_5_high_current_candle,
            min_15_high_current_candle, min_5_ema_50_current_candle, min_15_ema_50_current_candle,
            min_15_sma_100_current_candle, open, bc_close, rmp, pledge_benefit, lot_size, low_1yr,
            multiplier, pchange, pperchange, tick_size, fetched_at
        ) VALUES %s
        ON CONFLICT (isin) DO UPDATE SET
            sym = EXCLUDED.sym,
            disp_sym = EXCLUDED.disp_sym,
            exch = EXCLUDED.exch,
            inst = EXCLUDED.inst,
            seg = EXCLUDED.seg,
            seosym = EXCLUDED.seosym,
            mcap = EXCLUDED.mcap,
            mcapclass = EXCLUDED.mcapclass,
            pe = EXCLUDED.pe,
            div_yeild = EXCLUDED.div_yeild,
            revenue = EXCLUDED.revenue,
            year_1_revenue_growth = EXCLUDED.year_1_revenue_growth,
            net_profit_margin = EXCLUDED.net_profit_margin,
            yoy_last_qtrly_profit_growth = EXCLUDED.yoy_last_qtrly_profit_growth,
            ebidta_margin = EXCLUDED.ebidta_margin,
            volume = EXCLUDED.volume,
            price_perchng_1year = EXCLUDED.price_perchng_1year,
            price_perchng_3year = EXCLUDED.price_perchng_3year,
            price_perchng_5year = EXCLUDED.price_perchng_5year,
            ind_pe = EXCLUDED.ind_pe,
            pb = EXCLUDED.pb,
            eps = EXCLUDED.eps,
            day_sma_50_current_candle = EXCLUDED.day_sma_50_current_candle,
            day_sma_200_current_candle = EXCLUDED.day_sma_200_current_candle,
            day_rsi_14_current_candle = EXCLUDED.day_rsi_14_current_candle,
            roce = EXCLUDED.roce,
            ltp = EXCLUDED.ltp,
            roe = EXCLUDED.roe,
            rt_away_from_5_year_high = EXCLUDED.rt_away_from_5_year_high,
            rt_away_from_1_month_high = EXCLUDED.rt_away_from_1_month_high,
            high_5yr = EXCLUDED.high_5yr,
            high_3yr = EXCLUDED.high_3yr,
            high_1yr = EXCLUDED.high_1yr,
            high_1wk = EXCLUDED.high_1wk,
            price_perchng_1mon = EXCLUDED.price_perchng_1mon,
            price_perchng_1week = EXCLUDED.price_perchng_1week,
            price_perchng_3mon = EXCLUDED.price_perchng_3mon,
            yearly_earning_per_share = EXCLUDED.yearly_earning_per_share,
            ocf_growth_on_yr = EXCLUDED.ocf_growth_on_yr,
            year_1_cagr_eps_growth = EXCLUDED.year_1_cagr_eps_growth,
            net_change_in_cash = EXCLUDED.net_change_in_cash,
            free_cash_flow = EXCLUDED.free_cash_flow,
            price_perchng_2week = EXCLUDED.price_perchng_2week,
            day_bb_upper_sub_bb_lower = EXCLUDED.day_bb_upper_sub_bb_lower,
            day_atr_14_current_candle_mul_2 = EXCLUDED.day_atr_14_current_candle_mul_2,
            min_5_high_current_candle = EXCLUDED.min_5_high_current_candle,
            min_15_high_current_candle = EXCLUDED.min_15_high_current_candle,
            min_5_ema_50_current_candle = EXCLUDED.min_5_ema_50_current_candle,
            min_15_ema_50_current_candle = EXCLUDED.min_15_ema_50_current_candle,
            min_15_sma_100_current_candle = EXCLUDED.min_15_sma_100_current_candle,
            open = EXCLUDED.open,
            bc_close = EXCLUDED.bc_close,
            rmp = EXCLUDED.rmp,
            pledge_benefit = EXCLUDED.pledge_benefit,
            lot_size = EXCLUDED.lot_size,
            low_1yr = EXCLUDED.low_1yr,
            multiplier = EXCLUDED.multiplier,
            pchange = EXCLUDED.pchange,
            pperchange = EXCLUDED.pperchange,
            tick_size = EXCLUDED.tick_size,
            fetched_at = EXCLUDED.fetched_at;
    """
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            execute_values(cur, sql, records)
        conn.commit()
        logger.info(f"Upserted {len(records)} records into {T_CUSTOM_SCAN}.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to upsert records: {e}")
        raise e
    finally:
        conn.close()


# ---------------------------------------------------------------------------
# live_news table init + queries
# ---------------------------------------------------------------------------

def init_live_news_table(logger):
    """CREATE TABLE IF NOT EXISTS for live_news."""
    logger.info("Initializing live_news table...")
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(f"""
                CREATE TABLE IF NOT EXISTS {T_LIVE_NEWS} (
                    article_id BIGINT PRIMARY KEY,
                    title TEXT NOT NULL,
                    text TEXT,
                    overall_sentiment VARCHAR(50),
                    category VARCHAR(100),
                    sub_category VARCHAR(100),
                    publish_date TIMESTAMP WITH TIME ZONE,
                    stock_name VARCHAR(255),
                    isin_code VARCHAR(50),
                    sm_symbol VARCHAR(100),
                    nse_scrip_code INTEGER,
                    bse_scrip_code INTEGER,
                    seo_symbol VARCHAR(100),
                    display_symbol VARCHAR(100),
                    article_slug VARCHAR(255),
                    article_cat VARCHAR(100),
                    article_subcat VARCHAR(100),
                    metadata_json JSONB,
                    fetched_at TIMESTAMP WITH TIME ZONE NOT NULL
                );
            """)
        conn.commit()
        logger.info("live_news table initialized successfully.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to initialize table: {e}")
        raise
    finally:
        conn.close()


def load_existing_article_ids(conn, logger):
    """Return the set of article_id values already in live_news."""
    existing = set()
    with conn.cursor() as cur:
        try:
            cur.execute(f"SELECT article_id FROM {T_LIVE_NEWS};")
            for row in cur.fetchall():
                existing.add(row[0])
        except Exception as e:
            logger.warning(f"Could not load existing records: {e}")
    return existing


def insert_live_news_records(records, logger):
    """Upsert live_news tuples."""
    if not records:
        return

    sql = f"""
        INSERT INTO {T_LIVE_NEWS} (
            article_id, title, text, overall_sentiment, category, sub_category,
            publish_date, stock_name, isin_code, sm_symbol, nse_scrip_code,
            bse_scrip_code, seo_symbol, display_symbol, article_slug,
            article_cat, article_subcat, metadata_json, fetched_at
        ) VALUES %s
        ON CONFLICT (article_id) DO UPDATE SET
            title = EXCLUDED.title,
            text = EXCLUDED.text,
            overall_sentiment = EXCLUDED.overall_sentiment,
            category = EXCLUDED.category,
            sub_category = EXCLUDED.sub_category,
            publish_date = EXCLUDED.publish_date,
            stock_name = EXCLUDED.stock_name,
            isin_code = EXCLUDED.isin_code,
            sm_symbol = EXCLUDED.sm_symbol,
            nse_scrip_code = EXCLUDED.nse_scrip_code,
            bse_scrip_code = EXCLUDED.bse_scrip_code,
            seo_symbol = EXCLUDED.seo_symbol,
            display_symbol = EXCLUDED.display_symbol,
            article_slug = EXCLUDED.article_slug,
            article_cat = EXCLUDED.article_cat,
            article_subcat = EXCLUDED.article_subcat,
            metadata_json = EXCLUDED.metadata_json,
            fetched_at = EXCLUDED.fetched_at;
    """
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            execute_values(cur, sql, records)
        conn.commit()
        logger.info(f"Inserted {len(records)} records into {T_LIVE_NEWS}.")
    except Exception as e:
        conn.rollback()
        logger.error(f"Failed to insert news: {e}")
        raise
    finally:
        conn.close()
